File: programowanie zadania/Zakupy międzyplanetarne/zakupy_miedzyplanetarne.py
def zmiana_na_dziesietny(liczba, podstawa_poczatek):
    """Konwertuje liczbę z dowolnego systemu na dziesiętny."""
    wartosc_dziesietna = 0
    dlugosc = len(liczba)
    
    for i in range(dlugosc):
        cyfra = liczba[dlugosc - 1 - i]  # Iterowanie od końca
        if '0' <= cyfra <= '9':
            wartosc_cyfry = int(cyfra)
        else:
            wartosc_cyfry = ord(cyfra.upper()) - ord('A') + 10
        wartosc_dziesietna += wartosc_cyfry * (podstawa_poczatek ** i)
    return wartosc_dziesietna

def zmiana_z_dziesietnego(wartosc_dziesietna, podstawa_koniec):
    if wartosc_dziesietna == 0:
        return "0"
    
    wynik = ""
    while wartosc_dziesietna > 0:
        reszta = wartosc_dziesietna % podstawa_koniec
        if reszta < 10:
            wynik += str(reszta)
        else:
            wynik += chr(ord('A') + reszta - 10)
        wartosc_dziesietna //= podstawa_koniec

    wynik_odwrocony = ""
    for i in range(len(wynik) - 1, -1, -1):
        wynik_odwrocony += wynik[i]
    return wynik_odwrocony

def main():
    liczba = input("Podaj liczbe: ")
    podstawa_poczatek = int(input("Podaj podstawe systemu zrodlowego: "))
    podstawa_koniec = int(input("Podaj podstawe systemu docelowego: "))

    wartosc_dziesietna = zmiana_na_dziesietny(liczba, podstawa_poczatek)

    wynik = zmiana_z_dziesietnego(wartosc_dziesietna, podstawa_koniec)
    
    print(f"Liczba w systemie docelowym: {wynik}")

main()

File: programowanie zadania/Zakupy międzyplanetarne/zakupy_miedzyplanetarne.cpp
#include <iostream>
#include <string>
#include <cctype>
#include <cmath>

using namespace std;

int konwertujNaDziesietny(const string& liczba, int podstawa_poczatek) {
    int wartosc_dziesietna = 0;
    int dlugosc = liczba.length();

    for (int i = 0; i < dlugosc; ++i) {
        char cyfra = liczba[dlugosc - 1 - i];
        int wartosc_cyfry;
        if (isdigit(cyfra)) {
            wartosc_cyfry = cyfra - '0';
        } else {
            wartosc_cyfry = toupper(cyfra) - 'A' + 10;
        }
        wartosc_dziesietna += wartosc_cyfry * pow(podstawa_poczatek, i);
    }

    return wartosc_dziesietna;
}

string KonwertujZDziesietnego(int wartosc_dziesietna, int podstawa_koniec) {
    if (wartosc_dziesietna == 0) return "0";
    
    string wynik = "";
    while (wartosc_dziesietna > 0) {
        int reszta = wartosc_dziesietna % podstawa_koniec;
        if (reszta < 10) {
            wynik += (char)(reszta + '0');
        } else {
            wynik += (char)(reszta - 10 + 'A');
        }
        wartosc_dziesietna /= podstawa_koniec;
    }

    string wynik_odwrocony = "";
    for (int i = wynik.length() - 1; i >= 0; --i) {
        wynik_odwrocony += wynik[i];
    }
    return wynik_odwrocony;
}

int main() {
    string liczba;
    int podstawa_poczatek, podstawa_koniec;

    cout << "Podaj liczbe: ";
    cin >> liczba;
    cout << "Podaj podstawe systemu zrodlowego: ";
    cin >> podstawa_poczatek;
    cout << "Podaj podstawe systemu docelowego: ";
    cin >> podstawa_koniec;

    int wartosc_dziesietna = konwertujNaDziesietny(liczba, podstawa_poczatek);
    string wynik = KonwertujZDziesietnego(wartosc_dziesietna, podstawa_koniec);

    cout << "Liczba w systemie docelowym: " << wynik << endl;

    return 0;
}
